### PR TITLE
Deprecate lists for argument input on transpile()

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

Right now the support for the argument broadcasting with list inputs for various transpiler options on the transpile() function causes a significant performance overhead to support, primarily do to how we have to handle the multiple arguments across a parallel dispatch boundary. It also significantly increases the code complexity of the function to support more than one input for each argument (except circuits). The utility of doing this type of argument handling is quite limited since a similar result can be achieved with a for loop and would like be simpler for users to reason about. When weighing all these factors the best path forward is to just remove this functionality. This commit starts the process of removing this feature by marking it as deprecated. Once the deprecation cycle is complete we can greatly simplify the code in transpile and primarily replace it with a call to
generate_preset_pass_manager() and passmanager.run() (the only thing I think we'll have to handle out of band is faulty qubits defined in a BackendProperties for BackendV1).

Related to #7741

### Details and comments

### CODE REVIEW

File: qiskit/compiler/transpiler.py [hunk 1]
```diff
@@ -702,6 +704,41 @@ def _parse_transpile_args(
...
     }.items():
         if isinstance(value, list):
+            # This giant if statement is to resolve different argument
+            # handling special cases. For scheduling timing related parameters
+            # (timing_constraints and instruction_durations) are always
+            # expanded to a list. The intial_layout parameter can be a list
+            # object so it needs special handling to determin it's unique.
+            # This path is super buggy in general (outside of the warning) and
+            # since we're deprecating this it's better to just remove it than
+            # try to clean it up.
```
Comment (kevinhartman) [inline qiskit/compiler/transpiler.py:714]:
```suggestion
            # This giant if-statement detects deprecated use of argument
            # broadcasting. For arguments that previously supported broadcast
            # but were not themselves of type list (the majority), we simply warn
            # when the user provides a list. For the others, special handling is
            # required to disambiguate an expected value of type list from
            # an attempt to provide multiple values for broadcast. This path is
            # super buggy in general (outside of the warning) and since we're
            # deprecating this it's better to just remove it than try to clean it up.
```

The term "unique" seems context-specific to this method's impl, so I think it threw me off. I took a pass at rewording the original comment to be more accessible to someone less-familiar with `transpile`.

File: qiskit/compiler/transpiler.py [hunk 2]
```diff
@@ -88,22 +88,22 @@ def transpile(
 ) -> Union[QuantumCircuit, List[QuantumCircuit]]:
     """Transpile one or more circuits, according to some desired transpilation targets.
 
-    All arguments may be given as either a singleton or list. In case of a list,
-    the length must be equal to the number of circuits being transpiled.
+    .. deprecated:: 0.23.0
+
+        All arguments (except for ``circuits``) may be given as either a singleton or list. In
+        case of a list, the length must be equal to the number of circuits being transpiled. The
```
Comment (kevinhartman) [inline qiskit/compiler/transpiler.py:94]: > All arguments (except for ``circuits``) may be given as either a singleton or list. In case of a list, the length must be equal to the number of circuits being transpiled.

Is there a way to make it clear that this is a quote from the original documentation, and that it is actually the thing being deprecated? Not sure if there's a rule where we must include the deprecated documentation verbatim, but it could be helpful to reword this to make it clear that we're not describing the new functionality, but rather quoting what is no longer applicable.

**Review by kevinhartman — COMMENTED** [review]
LGTM, other than a few documentation-related requests.

File: qiskit/compiler/transpiler.py (hunk 2 above)
Comment (mtreinish) [reply qiskit/compiler/transpiler.py:94]: I can reword it, I was just being lazy and saw this block in the docs I just updated it to basically say "... and it's deprecated", I'll clean this up.

File: qiskit/compiler/transpiler.py [hunk 3]
```diff
@@ -88,22 +88,22 @@ def transpile(
...
+        support for using a list for an argument is deprecated and will be removed in the 0.25.0
+        release. If you need to use multiple values for an argument you can
+        use multiple :func:`~.transpile` calls (and potentially use :func:`~.parallel_map`
+        to leverage multiprocessing if needed).
```
Comment (kevinhartman) [inline qiskit/compiler/transpiler.py:98]:
```suggestion
        Previously, all arguments accepted lists of the same length as ``circuits``, which
        was used to specialize arguments for circuits at the corresponding indices.
        Support for using such argument lists is now deprecated and will be removed in
        the 0.25.0 release. If you need to use multiple values for an argument, you can
        use multiple :func:`~.transpile` calls (and potentially :func:`~.parallel_map` to
        leverage multiprocessing if needed).
```

**Review by kevinhartman — APPROVED** [review]

**Review by kevinhartman — APPROVED** [review]
